File: 프로그래머스/3/150367. 표현 가능한 이진트리/표현 가능한 이진트리.js
function solution(numbers) {
    var answer = [];
    let flag;
    
    for (let number of numbers) {
        
        let binaryStr = getBinary(number, "");
        
        let height = 1;
        
        while (true) {
            let tmp = Math.pow(2,height);
            
            if (binaryStr.length === 1) {
                break;
            }
            
            if (Math.floor(binaryStr.length / tmp) === 0) {
                break;
            }
            height++;
        }
        
        let pow = Math.pow(2, height) - 1;
        let diff = pow - binaryStr.length;
        
        for (let i = 0; i < diff; i++) {
            binaryStr = "0" + binaryStr;
        }
        
        let root = Math.floor(binaryStr.length / 2);
        
        let result = 1;
        global.flag = true;
        checkPerfectBinaryTree(root, binaryStr, height - 1);
        if (!global.flag) {
            result = 0;
        }
        answer.push(result);
        
    }
    return answer;
}

function checkPerfectBinaryTree(idx, binaryStr, height) {
    
    if (height === 0) {
        return;
    }
    
    let diff = Math.pow(2, height - 1);
    let left = idx - diff;
    let right = idx + diff;
    
    if (binaryStr[idx] === "0") {
        if (binaryStr[left] === "1" || binaryStr[right] === "1") {
           global.flag  = false;
        }
    }
    
    checkPerfectBinaryTree(left, binaryStr, height - 1);
    checkPerfectBinaryTree(right, binaryStr, height - 1);
    
}

function getBinary(number, result) {
    
    if (number === 0) {
        
        return result;
    }
    
    if (number % 2 === 0) {
        result = getBinary(Math.floor(number / 2), "0" + result);
    }
    
    if (number % 2 !== 0) {
        result = getBinary(Math.floor(number / 2), "1" + result);
    }
    
    return result;
}

File: 프로그래머스/3/150367. 표현 가능한 이진트리/표현 가능한 이진트리.java
import java.util.*;

class Solution {
    public int[] solution(long[] numbers) {
        int[] answer = new int[numbers.length];
        
        for (int i = 0; i < numbers.length; i++) {
            String binaryStr = convertToBinary(numbers[i], "");
            int height = getHeight(binaryStr);
            int diff =  (1 << height) - binaryStr.length() - 1;
            
            for (int j = 0; j < diff; j++) {
                binaryStr = "0" + binaryStr;
            }
            
            int root = binaryStr.length() / 2;
            int result = 1;
            if (!isPerfectBinaryTree(root, height - 1, binaryStr)) {
                result = 0;
            }
            answer[i] = result;
            
        }
        return answer;
    }
    
    private static int getHeight(String binaryStr) {
        
        int height = 1;
        int len = binaryStr.length();
        
        if (len == 1) {
            return height;
        }
        
        while (true) {
            if ((len / (1 << height)) == 0) {
                return height;
            }
            height++;
        }
    }
    
    private static String convertToBinary(long number, String result) {
        
        if (number == 0) {
            return result;
        }
        
        if ((number % 2) == 0) {
            result = convertToBinary((number / 2), "0" + result);
        }
        
        if ((number % 2) != 0) {
            result = convertToBinary((number / 2), "1" + result);
        }
        
        return result;
    }
    
    private static boolean isPerfectBinaryTree(int idx, int height, String binaryStr) {
        
        if (height == 0) {
            return true;
        }
        
        int diff =  (1 << height - 1);
        int left = idx - diff;
        int right = idx + diff;
        
        if (Objects.equals(binaryStr.charAt(idx), '0')) {
            if (Objects.equals(binaryStr.charAt(left), '1') || Objects.equals(binaryStr.charAt(right), '1')) {
                return false;
            }
        }
        
        return isPerfectBinaryTree(left, height - 1, binaryStr) && isPerfectBinaryTree(right, height - 1, binaryStr);
        
    }
}